apps/rush-lib/src/logic/deploy/DeployManager.ts

```diff
@@ -0,0 +1,588 @@
+    }
+  }
+
+  /**
+   * Recursively crawl the node_modules dependencies and collect the result in ISubdeploymentState.foldersToCopy.
+   */
+  private _collectFoldersRecursive(packageJsonFolderPath: string, subdemploymentState: ISubdeploymentState): void {
+    const packageJsonRealFolderPath: string = FileSystem.getRealPath(packageJsonFolderPath);
+
+    if (!subdemploymentState.foldersToCopy.has(packageJsonRealFolderPath)) {
+      subdemploymentState.foldersToCopy.add(packageJsonRealFolderPath);
+
+      const packageJson: IPackageJson = JsonFile.load(path.join(packageJsonRealFolderPath, 'package.json'));
+
+      // Union of keys from regular dependencies, peerDependencies, and optionalDependencies
+      const allDependencyNames: Set<string> = new Set<string>();
+      // Just the keys from optionalDependencies
+      const optionalDependencyNames: Set<string> = new Set<string>();
+
+      for (const name of Object.keys(packageJson.dependencies || {})) {
+        allDependencyNames.add(name);
+      }
+      if (this._deployScenarioJson.includeDevDependencies) {
+        for (const name of Object.keys(packageJson.devDependencies || {})) {
+          allDependencyNames.add(name);
+        }
+      }
+      for (const name of Object.keys(packageJson.peerDependencies || {})) {
+        allDependencyNames.add(name);
+        optionalDependencyNames.add(name); // consider peers optional, since they are so frequently broken
+      }
+      for (const name of Object.keys(packageJson.optionalDependencies || {})) {
+        allDependencyNames.add(name);
+        optionalDependencyNames.add(name);
+      }
+
+      for (const dependencyPackageName of allDependencyNames) {
+        // The "resolve" library models the Node.js require() API, which gives precedence to "core" system modules
+        // over an NPM package with the same name.  But we are traversing package.json dependencies, which refer
+        // to system modules.  Appending a "/" forces require() to load the NPM package.
+        const resolveSuffix: string = dependencyPackageName + resolve.isCore(dependencyPackageName) ? '/' : '';
+
+        const resolvedDependency: string = resolve.sync(dependencyPackageName + resolveSuffix, {
+          basedir: packageJsonRealFolderPath,
+          preserveSymlinks: false,
+          packageFilter: (pkg, dir) => {
+            // point "main" at a file that is guaranteed to exist
+            // This helps resolve packages such as @types/node that have no entry point
+            pkg.main = "./package.json";
+            return pkg;
+          },
+          realpathSync: (filePath) => {
+            try {
+              const resolvedPath: string = require("fs").realpathSync(filePath);
+
+              subdemploymentState.symlinkAnalyzer.analyzePath(filePath);
+              return resolvedPath;
+            } catch (realpathErr) {
+              if (realpathErr.code !== "ENOENT") {
```

Comment:
Use the `FileSystem.isFileNotExistError` function?

Reply:
Again I just copied the documented incantation verbatim.